lib/kafka_producer.js

```diff
@@ -262,16 +296,35 @@ KafkaProducer.prototype.getWholeMsg = function getWholeMsg(topic, message, timeS
 
 KafkaProducer.prototype.close = function close(callback) {
     var self = this;
+
+    self.enable = false;
+    self.closing = true;
+    self.closeCallback = callback;
+
+    if (self.flushInterval) {
+        clearInterval(self.flushInterval);
+    }
+
     if (self.batching) {
-        self.flushEntireCache(callback);
+        self.flushEntireCache();
     }
-    self.enable = false;
+
+    self._tryClose();
```

Comment:
If the close fails on ```self.pendingWrites !== 0``` there is no retry or error returned to the callback so it will block the shutdown because the shutdown only finishes when the callback is ran.